### PR TITLE
Singapore morning.

### PR DESCRIPTION
**Pull request by sunjincheng121**
## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*

## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Expression.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package org.apache.flink.table.expressions;
+
+import java.util.List;
+
+/**
+ * Expression.
+ */
+public abstract class Expression {
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Expression.java]: +1

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Alias.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+	public static Alias apply(Expression child, String name) {
+		return new Alias(child, name);
+	}
+
+	private Expression child;
+	private String name;
+	private List<String> extraNames;
+
+	public Alias(Expression child, String name, List<String> extraNames) {
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Alias.java]: +1

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/expressions/PlannerExpression.scala [hunk 3]
```diff
@@ -72,17 +72,17 @@ abstract class Expression extends TreeNode[Expression] {
...
 
-abstract class UnaryExpression extends Expression {
-  private[flink] def child: Expression
+abstract class PlannerUnaryExpression extends PlannerExpression {
+  private[flink] def child: PlannerExpression
   private[flink] def children = Seq(child)
 }
 
-abstract class LeafExpression extends Expression {
+abstract class PlannerLeafExpression extends PlannerExpression {
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/expressions/PlannerExpression.scala]: +1

**Comment by hequn8128** [comment]
注重看了下jira上timo的回复，一些想法如下（不一定100%正确，大家一起讨论下）：
- 从`FunctionDefinition`角度，不用区分builtin还是user defined。只保存function的name和type就行。对应类的结构是，基类`FunctionDefinition`（包含变量name），子类`AggFunctionDefinition`, `ScalarFunctionDefinition`, `TableFunctionDefinition `(没有变量)。`BuildInFunctionDefinition`这个类感觉可以去了？`FunctionDefinitions`里的所有内置function，可以用`AggFunctionDefinition`, `ScalarFunctionDefinition`, `TableFunctionDefinition `表示。
这里需要解决一个问题：对应于scala用户，怎么保存`UDAGGExpression`里的resultTypeInfo和accTypeInfo？有一个解法是类似于java，先自动register用户的udf，并生成register的name，后面再根据name来找到function并替换，替换逻辑参考`replaceAggFunctionCall`。

- `FunctionDefinitions`需要实现`getDefinitions()`，类似于`SqlStdOperatorTable`里`getOperatorList()`的反射实现（继承自`ReflectiveSqlOperatorTable`）。

- Expression添加`accept()`方法，然后目前的`ScalaExpressionParser`用`ExpressionVisitor`来实现。可能会好些，比如可以为test case编写一个visitor来打印api树的结构, 等等。

代码细节，明天我再看下

**Comment by WeiZhong94** [comment]
@hequn8128 
1. 现在这一版实现是在timo的第二个评论之前写出来的..也就是参照他的第一个评论实现出来的代码. 那一版里面他区分出来了一个UserDefinedFunctionDefinition, 所以按照他的意思我把BulitInFunctionDefinition也给实现出来了. 另外, 我感觉就算概念上不分udf和built-in, 实现上总会有一个地方要把它们区分出来的. 即使全用* `AggFunctionDefinition`, `ScalarFunctionDefinition`, `TableFunctionDefinition` *, 里面是不是也得找个办法区分udf和built-in? udf就从Definition里直接拿用户的函数, built-in就去function catalog里面 lookup.  如果要不区分, 那就必须自动register用户的udf. 这么做的话, 在实现的某个步骤里应该是要用到tEnv的. 感觉这么做为了概念上的干净, 其实把实现复杂化了, 而且引入了不必要的依赖, 隐隐感觉有点不妥..
2. 没实现这个的主要原因是发现暂时好像用不上这个方法..
3. scala的match case本来就是用来替代java的visitor 模式的, 更加简洁可读性也更好. 而且planner 短期也不可能去scala, 所以就用scala的match case来实现了..

**Comment by WeiZhong94** [comment]
这些问题先讨论一下我再修改吧..我现在在把TableFunction的Definition加上, 刚刚看了一下发现现在TableFunciton的隐式转换依赖了LogicalNode了, 这里估计要重构一下..还有就是之前说的各个测试代码的修改, 我先做这些部分的内容吧.

**Comment by hequn8128** [comment]
@WeiZhong94 是的，目前发现统一没有太大好处，实现会复杂很多，我今天提交的diff里没有统一

**Pull request closed without merging**